### PR TITLE
[WIP] Allowing `let` in rhs

### PR DESCRIPTION
**Pull request by gabrielhdt**
See #215

### CODE REVIEW

File: src/scope.ml [hunk 1]
```diff
@@ -367,12 +367,13 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
     | (P_Prod(_,_)     , M_Patt   ) ->
         fatal t.pos "Dependent products are not allowed in a pattern."
     | (P_Prod(xs,b)    , _                ) -> scope_binder _Prod env xs b
-    | (P_LLet(x,xs,t,u), M_Term(_)        ) ->
+    | (P_LLet(x,xs,t,u), M_Term(_)        )
+    | (P_LLet(x,xs,t,u), M_RHS(_)         ) ->
         (* “let x = t in u” is desugared as “(λx.u) t” (for now). *)
         let t = scope env (if xs = [] then t else Pos.none (P_Abst(xs,t))) in
         _Appl (scope env (Pos.none (P_Abst([([Some x],None,false)], u)))) t
     | (P_LLet(_,_,_,_) , _                ) ->
```
Comment (fblanqui) [inline src/scope.ml]: please change to `     | (P_LLet(_,_,_,_) , M_LHS(_)                ) ->`

**Comment by fblanqui** [comment]
Please add some test file.

**Comment by gabrielhdt** [comment]
I face a strange behaviour, the following code typechecks
```
rule times (frac &a &b) (frac &c &d) →
  frac
   (N.times &a &c)
   (N.nznatc
    (N.times (fst N.not_zero &b) (fst N.not_zero &d))
    (N.prod_not_zero
     (fst N.not_zero &b)
     (fst N.not_zero &d)
     (snd N.not_zero &b)
     (snd N.not_zero &d)))
```
but this one does not,
```
rule times (frac &a &b) (frac &c &d) →
  let bv ≔ fst N.not_zero &b in
  frac
   (N.times &a &c)
   (N.nznatc
    (N.times bv (fst N.not_zero &d))
    (N.prod_not_zero
     bv
     (fst N.not_zero &d)
     (snd N.not_zero &b)
     (snd N.not_zero &d)))
```
I'll try to get a minimal example

File: src/basics.ml [hunk 2]
```diff
@@ -100,6 +100,10 @@ let eq : term -> term -> bool = fun a b -> a == b ||
     | (Prod(a1,b1), Prod(a2,b2))
     | (Abst(a1,b1), Abst(a2,b2)) -> let (_, b1, b2) = Bindlib.unbind2 b1 b2 in
                                     eq ((a1,a2)::(b1,b2)::l)
+    | (LLet(t1,a1,u1), LLet(t2,a2,u2)) ->
+        let (_, u1, u2) = Bindlib.unbind2 u1 u2 in
+        (* FIXME or equality of substituted? *)
```
Comment (rlepigre) [inline src/basics.ml]: No, what you are doing is correct. This is not conversion, but syntactic equality (up to α-equivalence).

**Comment by gabrielhdt** [comment]
This PR for now tries to implement `let` the light way, that is, during type inference, as soon as a `LLet(t,a,u)` is encountered, we substitute `t` in `u` and continue inference.

This process raises issues regarding meta variables, their free variables (e.g. `?0[x,y]`) can be substituted by the `let` rule, yielding non valid variables since they are not applied to distinct bound variables (but some term).

The other way
------------------
the branch [`heavy-let`](https://github.com/gabrielhdt/lambdapi/tree/heavy-let) implements what is done in other proof assistants, that is, contexts with values and unification with context.

**Comment by rlepigre** [comment]
> This process raises issues regarding meta variables, their free variables (e.g. `?0[x,y]`) can be substituted by the `let` rule, yielding non valid variables since they are not applied to distinct bound variables (but some term).

I don't think that's a problem, we just need to use the equivalence `?0[x,y] = (λx.λy.?0[x,y]) x y`. (Inserting the minimal number of β-expansions to avoid as many higher-order unification problems as possible.)

**Pull request closed without merging**